org-ai-openai-image-query.el

```diff
@@ -0,0 +1,124 @@
+
+(defvar org-ai-openai-image-query-endpoint "https://api.openai.com/v1/chat/completions"
+  "Endpoint for querying images with OpenAI.")
+
+(defun org-ai--send-image-query (image-content question callback)
+  "Send IMAGE-CONTENT and QUESTION to the OpenAI API.
+Calls CALLBACK with the response."
+  (let ((url-request-extra-headers `(("Authorization" . ,(string-join `("Bearer" ,(org-ai--openai-get-token)) " "))
+                                     ("Content-Type" . "application/json")))
+        (url-request-method "POST")
+        (url-request-data (json-encode `(("model" . ,org-ai-image-query-model)
+                                         ("messages" . [(("role" . "user")
+                                                         ("content" . [((type . "text") (text . ,question))
+                                                                       ,image-content]))])
+                                         ("max_tokens" . ,(or org-ai-default-max-tokens 300))))))
+    (url-retrieve
+     org-ai-openai-image-query-endpoint
+     (lambda (_status)
+       (goto-char url-http-end-of-headers)
+       (let ((response (json-read)))
+         (if response
+             (funcall callback response)
+           (message "Failed to get a valid response from OpenAI API.")))))))
+
+(defun org-ai--send-base64-image-query (base64-image question callback)
+  "Send a BASE64-IMAGE and QUESTION to the OpenAI API.
+Calls CALLBACK with the response."
+  (org-ai--send-image-query
+   `((type . "image_url")
+     (image_url . (("url" . ,(concat "data:image/jpeg;base64," base64-image)))))
+   question callback))
+
+(defun org-ai--send-url-image-query (image-url question callback)
+  "Send an IMAGE-URL and QUESTION to the OpenAI API.
+Calls CALLBACK with the response."
+  (org-ai--send-image-query
+   `((type . "image_url")
+     (image_url . (("url" . ,image-url))))
+   question callback))
+
+(defun org-ai--handle-openai-response (response)
+  "Handle the RESPONSE from OpenAI API."
+  (let* ((choices (alist-get 'choices response))
+         ;; Ensure choices is a vector and has at least one element
+         (choice (and (vectorp choices)
+                      (> (length choices) 0)
+                      (aref choices 0)))
+         (message_ (alist-get 'message choice))
+         (content (alist-get 'content message_))
+         (output-buffer (get-buffer-create (or org-ai-query-image-file "*org-ai-output*"))))
+    (if content
+        (org-ai-prompt--insert output-buffer content t)
+      (error "Content not found in the response"))))
+
+(defun org-ai--get-image-path-or-url ()
+  "Prompt the user for a non-empty image path or URL."
+  (let ((default (or (thing-at-point 'url)
+                     (thing-at-point 'filename)))
+        (input ""))
+    (while (string-empty-p input)
+      (setq input
+            (read-string (if default
+                             (format "Enter image URL or file path (default %s): " default)
```

Comment:
consider adding "URL must start with `http?s://`

Reply:
feels bulky to add more text, and people probably already know that urls must start with http[s]:// in emacs right?